src/dbnode/persist/types.go

```diff
@@ -21,20 +21,133 @@
 package persist
 
 import (
+	"errors"
 	"fmt"
 	"time"
 
 	"github.com/m3db/m3/src/dbnode/namespace"
 	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/m3ninx/doc"
 	"github.com/m3db/m3/src/m3ninx/index/segment"
 	idxpersist "github.com/m3db/m3/src/m3ninx/persist"
 	"github.com/m3db/m3/src/x/ident"
 
 	"github.com/pborman/uuid"
 )
 
+var (
+	errReuseableTagIteratorRequired = errors.New("reuseable tags iterator is required")
+)
+
+// Metadata is metadata for a time series, it can
+// have several underlying sources.
+type Metadata struct {
+	metadata doc.Document
+	id       ident.ID
+	tags     ident.Tags
+	tagsIter ident.TagIterator
+	opts     MetadataOptions
+}
+
+// MetadataOptions is options to use when creating metadata.
+type MetadataOptions struct {
+	FinalizeID          bool
+	FinalizeTags        bool
+	FinalizeTagIterator bool
+}
+
+// NewMetadata returns a new metadata struct from series metadata.
+func NewMetadata(metadata doc.Document) Metadata {
+	return Metadata{metadata: metadata}
+}
+
+// NewMetadataFromIDAndTags returns a new metadata struct from
+// explicit ID and tags.
+func NewMetadataFromIDAndTags(
+	id ident.ID,
+	tags ident.Tags,
+	opts MetadataOptions,
+) Metadata {
+	return Metadata{
+		id:   id,
+		tags: tags,
+		opts: opts,
+	}
+}
+
+// NewMetadataFromIDAndTagIterator returns a new metadata struct from
+// explicit ID and tag iterator.
+func NewMetadataFromIDAndTagIterator(
+	id ident.ID,
+	tagsIter ident.TagIterator,
+	opts MetadataOptions,
+) Metadata {
+	return Metadata{
+		id:       id,
+		tagsIter: tagsIter,
+		opts:     opts,
+	}
+}
+
+// BytesID returns the bytes ID of the series.
+func (m Metadata) BytesID() []byte {
+	if m.id != nil {
+		return m.id.Bytes()
+	}
+	return m.metadata.ID
+}
+
+// MetadataTagIteratorOptions are options required to be passed
+// to metadata TagIterator to retrieve a tag iterator.
+type MetadataTagIteratorOptions struct {
+	ReuseableTagsIterator ident.TagsIterator
+}
+
+// TagIterator returns a tag iterator for the series,
+// returning a direct ref to a provided tag iterator
+// or using the reuseable tag iterator provided by the
+// callsite if it needs to iterate over tags or fields.
+func (m Metadata) TagIterator(
+	opts MetadataTagIteratorOptions,
+) (ident.TagIterator, error) {
+	reuseable := opts.ReuseableTagsIterator
+	if reuseable == nil {
+		// Always check to make sure callsites won't
+		// get a bad allocation pattern of having
+		// to create one here inline if the metadata
+		// they are passing in suddenly changes from
+		// tagsIter to tags or fields with metadata.
+		return nil, errReuseableTagIteratorRequired
```

Comment:
If we always require a reusable tag iterator should we just make it a required argument?

Reply:
Will get rid of the options.